### ISSUE TITLE
[BUG] Delegating take_focus focuses all elements

### ISSUE THREAD

**Issue by tpmoney** [gyscos/cursive#660]
I'm trying to build a custom view that has a couple focusable elements. I'm reusing Button elements as part of this. In the "take_focus" function, I'm determining which button should take focus and then passing the "take_focus" call onto that button. When the view is used, all the buttons appear to take focus instead of just the chosen button.

### test_focus.rs
```rust test_focus.rs
use cursive::{
    event::EventResult,
    reexports::log::info,
    views::{Button, Dialog},
    View, XY,
};

pub struct TestFocus {
    button1: Button,
    button2: Button,
    button3: Button,
}

impl View for TestFocus {
    fn draw(&self, printer: &cursive::Printer) {
        self.button1.draw(printer);
        self.button2.draw(&printer.offset((0, 1)));
        self.button3.draw(&printer.offset((0, 2)))
    }

    fn required_size(&mut self, constraint: XY<usize>) -> XY<usize> {
        self.button1
            .required_size(constraint)
            .saturating_add(self.button2.required_size(constraint))
            .saturating_add(self.button2.required_size(constraint))
    }

    fn take_focus(
        &mut self,
        source: cursive::direction::Direction,
    ) -> Result<cursive::event::EventResult, cursive::view::CannotFocus> {
        info!("Taking focus!");
        self.button1.take_focus(source)
    }

    fn focus_view(
        &mut self,
        _: &cursive::view::Selector<'_>,
    ) -> Result<cursive::event::EventResult, cursive::view::ViewNotFound> {
        info!("Called focus_view");
        Ok(EventResult::consumed())
    }
}

impl Default for TestFocus {
    fn default() -> Self {
        Self::new()
    }
}

impl TestFocus {
    pub fn new() -> Self {
        TestFocus {
            button1: Button::new("Button 1", |term| term.add_layer(Dialog::info("Button 1"))),
            button2: Button::new("Button 2", |term| term.add_layer(Dialog::info("Button 2"))),
            button3: Button::new("Button 3", |term| term.add_layer(Dialog::info("Button 3"))),
        }
    }
}
```

### main
```rust
let mut term = Cursive::new();
cursive::logger::init();
cursive::reexports::log::set_max_level(LevelFilter::Info);
term.add_layer(
        Panel::new(
            LinearLayout::vertical()
                .child(EditView::new().content("Hello Test Focus"))
                .child(TestFocus::new()),
        )
        .fixed_width(40)
        .fixed_height(40),
    );
term.run();
```

I would expect with this code for `Button 1` to have focus highlight and buttons 2/3 to remain unfocused. Instead when the `TestFocus` component is focused, all the buttons show focus. Additionally only 1 "Taking focus!" log message is logged and no "Called focus_view" messages, so it doesn't appear that multiple focus events are being fired accidentally.

**Screenshots**
<img width="294" alt="Screen Shot 2022-07-10 at 10 49 43" src="https://user-images.githubusercontent.com/4555240/178149921-14a244f5-1042-4087-8cb7-eaeef007df77.png">

<img width="313" alt="Screen Shot 2022-07-10 at 10 50 03" src="https://user-images.githubusercontent.com/4555240/178149932-af0610db-e382-4633-8808-09e021849eed.png">

**Environment**
* Operating System: macOS 12.4
* Backend used: ncurses
* Current locale:
```
LANG="en_US.UTF-8"
LC_COLLATE="en_US.UTF-8"
LC_CTYPE="en_US.UTF-8"
LC_MESSAGES="en_US.UTF-8"
LC_MONETARY="en_US.UTF-8"
LC_NUMERIC="en_US.UTF-8"
LC_TIME="en_US.UTF-8"
LC_ALL=
```
* Cursive version: 0.18.0

**Comment by tpmoney** [comment]
This apparently also affected the FixedLayout view as well. If you replace the TestFocus code with the following code, the result is the same for the focus behavior. One difference is in the custom view above, hitting "return" doesn't trigger any of the buttons, which I assume is because I need to handle events and delegate. In this code, "Button 1" will trigger but not the others. Pressing "tab" and then "return" will trigger button 2, and doing again will trigger button 3. So it seems like actual focus is still on the expected button, but the display code is treating them all as focused.

```rust
use cursive::{
    view::ViewWrapper,
    views::{Button, Dialog, FixedLayout},
    wrap_impl, Rect,
};

pub struct TestFocus {
    layout: FixedLayout,
}

impl ViewWrapper for TestFocus {
    wrap_impl!(self.layout: FixedLayout);
}

impl Default for TestFocus {
    fn default() -> Self {
        Self::new()
    }
}

impl TestFocus {
    pub fn new() -> Self {
        TestFocus {
            layout: FixedLayout::new()
                .child(
                    Rect::from_size((0usize, 0), (15usize, 1)),
                    Button::new("Button 1", |term| term.add_layer(Dialog::info("Button 1"))),
                )
                .child(
                    Rect::from_size((0usize, 1), (15usize, 1)),
                    Button::new("Button 2", |term| term.add_layer(Dialog::info("Button 2"))),
                )
                .child(
                    Rect::from_size((0usize, 2), (15usize, 1)),
                    Button::new("Button 3", |term| term.add_layer(Dialog::info("Button 3"))),
                ),
        }
    }
}
```

**Comment by gyscos** [comment]
Hi, and thanks for the report!

One thing to keep in mind is that in Cursive, views don't store in their state whether they are focused. Instead, the _parent_ view is responsible for keeping track of which of its children is focused. As a matter of fact, a view group is even free to have more than one child "focused" at a time! (Though the built-in view groups only keep one child focused at a time.)

Only when being drawn do the views know if they are focused (so they can change some highlighting colors). [`Printer` has a `focused` field](https://docs.rs/cursive/latest/cursive/struct.Printer.html#structfield.focused) to indicate that.

In your first `TestFocus` example, in the `draw` method, you directly pass the `printer` to the children views. Instead, you should use [`Printer::focused()`](https://docs.rs/cursive/latest/cursive/struct.Printer.html#method.focused) to "disable" focus for all but one of the children.
To do that, you will probably need to keep track of which view is focused. For example, [`LinearLayout` uses a `focus` field for that](https://docs.rs/cursive_core/0.3.1/src/cursive_core/views/linear_layout.rs.html#478).

Regarding your second example with `FixedLayout`... it's a bug indeed! [You can check the draw method](https://docs.rs/cursive_core/0.3.1/src/cursive_core/views/fixed_layout.rs.html#276): it does not change the focus there (but totally should).

**Comment by tpmoney** [comment]
Ah, I thought I might just be doing something wrong but was having a hard time figuring out what it was. Thanks. I'll leave this open to track the Fixed Layout bug?

**Comment by gyscos** [comment]
Sure! It should be fixed on the latest commit, will make a release shortly.

**Comment by gyscos** [comment]
Should be released as part of cursive-core 0.3.4.
